### DISCUSSION TITLE
Collecting Errors Raised by Custom Validators, Before Raising

### DISCUSSION THREAD

**Discussion by JKFSOM** [pydantic/pydantic#7470] [Question] [answered]
In Pydantic v1 it was possible to collect `ValueError`s, then raise them as a list of the `ErrorWrapper` class. This allowed (i) validation to be carried out on all elements, without stopping at the first failure; (ii) to easily output multiple errors in a style similar to that raised by the built-in Pydantic model validation - not concatenated. 

Is something similar possible in Pydantic v2?

- SO post containing the approach outlined above: https://stackoverflow.com/a/72003724

**Comment by samuelcolvin** [comment, marked as answer]
Yes, you can build validation errors manually - https://github.com/pydantic/pydantic-core/blob/f6b14cce22b1539727ed0e0a47d2c680c79c1692/python/pydantic_core/_pydantic_core.pyi#L680

**Reply by JKFSOM** [reply]
Thanks for your help @samuelcolvin.

**Reply by KeynesYouDigIt** [reply]
Could we get some more complex / stable examples of using this that are "canon"? I think doing so would close out https://github.com/pydantic/pydantic-core/issues/963 . I see a few interesting ideas below.... is there one the maintainers would sign off on as "best practice"?

**Comment by cryzed** [comment]
I think the intent behind the question was to find some way to validate a model exhaustively, i.e. not stop after the first found error for some field, but continue validation with all other known validators of the model, including: validators that are part of any annotated types used in the model, methods decorated with `pydantic.field_validator`, `pydantic.model_validator` etc.

The use-case behind this would be to find _all_ issues with the model's data at once, allowing for example to highlight _all_ erroneous fields to the end-user in a dialogue. This is something I'd like to do myself currently too: validate a model exhaustively and then prompt my end-users to please correct all highlighted fields (but I want to highlight all invalid fields at once, not field-by-field!).

I don't see how building and raising `pydantic_core.ValidationError`s manually would allow us to easily do this, without resorting to creating a single model-validator method that does all validation at once and collates it -- but that would be giving up a big chunk of pydantic's niceties (i.e. the nice separation of concerns with the `pydantic.field_validators`, the fact that we can access `loc` as part of a `pydantic.ErrorDetail` to see which field/value caused the issue, etc.

I think this is a common-enough use-case that it makes sense to support validation like this directly in pydantic, unless I am "using pydantic wrongly" and this is not what its intended for (i.e. something close to form-data validaton).

The current work-around I am using, to collect as many invalid fields and their errors from my own models as possible, is to validate my model in a loop, replacing the values for failing fields with their default values specified in the model, until the list of errors stops changing.

**Comment by JKFSOM** [comment]
Thanks for your comment @cryzed. I believe I've achieved what you're looking for.

It requires two things: (i) a `model_validator` with `mode='wrap'`; (ii) `validation_error. from_exception_data()` to collect and later raise exceptions.

(clearly improvements could be made to the following code - the use of enums for `height_type` to start, but hopefully it acts as a guide.)

```python
from pydantic import BaseModel, model_validator, ValidationError
from pydantic_core import InitErrorDetails

class ListingModel(BaseModel):
   gender: str
   height_type: str
   height_cms: float = Field(default=None)
   height_hands: float = Field(default=None)

   @model_validator(mode='wrap')
   def check_height_type(self, handler) -> Any:
      validation_errors = []

      # handle custom validation, before validating the model itself
      height_type = self.get("height_type")
      height_cms = self.get("height_cms")

      if height_type == "cms":
         if height_cms is None:
            validation_errors.append(
               InitErrorDetails(
                  type="missing",
                  loc=("height_cms",),
                  input="height_cms",
                  ctx={},
               )
            )
         
      validated_self = None
      try:
         # validate the model
         validated_self = handler(self)
      except ValidationError as e:
         validation_errors.extend(e.errors())

      if validation_errors:
         raise ValidationError.from_exception_data(title=self.__class__.__name__, line_errors=validation_errors)

      return validated_self
```

When `height_type` is set to `cms` but a `height_cms` value isn't provided, and `gender` is omitted, this should return something like the following:
```
{
	"detail": [
		{
			"type": "missing",
			"loc": [
				"body",
				"height_cms"
			],
			"msg": "Field required",
			"input": null,
			"ctx": {},
			"url": "https://errors.pydantic.dev/2.3/v/missing"
		},
		{
			"type": "missing",
			"loc": [
				"body",
				"gender"
			],
			"msg": "Field required",
			"input": {
				"height_type": "cms"
			},
			"url": "https://errors.pydantic.dev/2.3/v/missing"
		}
	]
}
```

**Reply by cryzed** [reply]
It's great that you got it working for your use-case, but this looks like I'd have to change how I structure and write validators majorly (something I think should be pydantic's task to make work nicely) -- what I was alluding to in my first comment about "giving up a big chunk of pydantic's niceties".

It just make sense to me to have something like `BaseModel.model_validate(obj, exhaust=True)` or `BaseModel.model_validate_json(obj, exhaust=True)` to collect all possible errors that can be collected from the model.

**Reply by cryzed** [reply]
Below is one of the very ugly/hacky prototype ways I came up with to fully validate one of my models the way I am intending. Keep in mind, that my models have quite a few fields, and most of them are optional (`None`-default value). The code doesn't handle all validator-types (mostly wrap-validators and field-validators with non-default signatures), and I am sure there are some things I misunderstood about the pydantic internals, but at least in my tests it seems to do what I want.

EDIT: I removed the code snippet because I realized that this is much more complicated to do exhaustively (and correctly) than I initially thought.

**Reply by forrest-bajbek** [reply]
This was really helpful. Thank you.

**Comment by forrest-bajbek** [comment]
I second the sentiment that PydanticV2 should have some mechanism to, in a single method, append multiple instances of `ValueError` to the final `PydanticValidationError`. While support for isolated unit tests on each field is fantastic, support for integration testing on a model is terrible without this feature. `ErrorWrapper` used to solve this problem.

**Reply by forrest-bajbek** [reply]
After working through the problem, I realize `pydantic.ValidationError.from_exception_data` solves this problem, though [the documentation is lacking](https://docs.pydantic.dev/latest/api/pydantic_core/#pydantic_core._pydantic_core.ValidationError.from_exception_data).

For anyone who finds this thread, here is a helpful, working example:
```python
from enum import Enum
from typing import Optional

from pydantic import BaseModel, Field, ValidationError
from pydantic_core import InitErrorDetails, PydanticCustomError

class Airport(Enum):
    FLL = "Fort Lauderdale, FL"
    JAX = "Jacksonville, FL"
    MIA = "Miami, FL"

class AirlineTicket(BaseModel):
    leaving_source_airport: Airport = Field()
    leaving_destination_airport: Airport = Field()
    is_round_trip: bool = Field(default=False)
    returning_source_airport: Optional[Airport] = Field(default=None)
    returning_destination_airport: Optional[Airport] = Field(default=None)

    def model_post_init(self, __context):
        validation_errors: list[InitErrorDetails] = list()

        if self.is_round_trip is True:
            if self.returning_source_airport:
                validation_errors.append(
                    InitErrorDetails(
                        type=PydanticCustomError(
                            "integration_test",
                            "When 'is_round_trip` is True, returning_source_airport should be left empty.",
                        ),
                        loc=("returning_source_airport",),
                        input=self.returning_source_airport,
                        ctx={},
                    )
                )
            if self.returning_destination_airport:
                validation_errors.append(
                    InitErrorDetails(
                        type=PydanticCustomError(
                            "integration_test",
                            "When 'is_round_trip` is True, returning_destination_airport should be left empty.",
                        ),
                        loc=("returning_destination_airport",),
                        input=self.returning_destination_airport,
                        ctx={},
                    )
                )
        elif self.is_round_trip is False:
            if not self.returning_source_airport:
                validation_errors.append(
                    InitErrorDetails(
                        type=PydanticCustomError(
                            "integration_test",
                            "When 'is_round_trip` is False, returning_source_airport is required.",
                        ),
                        loc=("returning_source_airport",),
                        input=self.returning_source_airport,
                        ctx={},
                    )
                )
            if not self.returning_destination_airport:
                validation_errors.append(
                    InitErrorDetails(
                        type=PydanticCustomError(
                            "integration_test",
                            "When 'is_round_trip` is False, returning_destination_airport is required.",
                        ),
                        loc=("returning_destination_airport",),
                        input=self.returning_destination_airport,
                        ctx={},
                    )
                )

        if validation_errors:
            raise ValidationError.from_exception_data(
                title=self.__class__.__name__,
                line_errors=validation_errors,
            )

try:
    ticket = AirlineTicket(
        leaving_source_airport=Airport.FLL,
        leaving_destination_airport=Airport.JAX,
        is_round_trip=True,
        returning_source_airport=Airport.JAX,
        returning_destination_airport=Airport.MIA,
    )
except ValidationError as e:
    for err in e.errors():
        print(err)
# {'type': 'integration_test', 'loc': ('returning_source_airport',), 'msg': "When 'is_round_trip` is True, returning_source_airport should be left empty.", 'input': <Airport.JAX: 'Jacksonville, FL'>}
# {'type': 'integration_test', 'loc': ('returning_destination_airport',), 'msg': "When 'is_round_trip` is True, returning_destination_airport should be left empty.", 'input': <Airport.MIA: 'Miami, FL'>}

try:
    ticket = AirlineTicket(
        leaving_source_airport=Airport.FLL,
        leaving_destination_airport=Airport.JAX,
    )
except ValidationError as e:
    for err in e.errors():
        print(err)
# {'type': 'integration_test', 'loc': ('returning_source_airport',), 'msg': "When 'is_round_trip` is False, returning_source_airport is required.", 'input': None}
# {'type': 'integration_test', 'loc': ('returning_destination_airport',), 'msg': "When 'is_round_trip` is False, returning_destination_airport is required.", 'input': None}
```

**Comment by salomon-smekecohen** [comment]
Look I dont think this will work for everyone. In particular it ignores plugins and config completely. It is also slow and i didnt even try before validators. I wrote it as POC for my use case. May the concept prove useful.

```python
from __future__ import annotations

from typing import Any
from typing import Dict
from typing import cast

from pydantic import BaseModel
from pydantic import ValidationError
from pydantic_core import ErrorDetails
from pydantic_core import SchemaValidator


def prune(schema: dict) -> None:
    needle: dict | None = schema
    stack = [schema]
    deepest = 0

    while needle:
        if needle := needle.get("schema"):
            stack.append(needle)
            if "fields" in needle:
                deepest = len(stack)

    stack = stack[:deepest]
    if not deepest or len(stack) < 3:
        raise IndexError

    stack[-3]["schema"] = stack[-1]


class ExhaustiveModel(BaseModel):
    def __init__(__pydantic_self__, **data: dict[str, Any]) -> None:
        # Obtain a copy of the pydantic schema.
        schema = cast(Dict[str, Any], __pydantic_self__.__pydantic_validator__.__reduce__()[1][0])
        errors: dict[str, ErrorDetails] = {}

        try:
            while True:
                try:
                    # Warning: mutation.
                    SchemaValidator(schema).validate_python(data, self_instance=__pydantic_self__)
                    break
                except ValidationError as err:
                    errors.update({str(e): e for e in err.errors()})

                prune(schema)
        except IndexError:
            pass

        if errors:
            raise ValidationError.from_exception_data(title=__pydantic_self__.__class__.__name__, line_errors=[*errors.values()])

        super().__init__(**data)
```

```python
from pydantic import model_validator, field_validator
class SomeModel(ExhaustiveModel):
    @model_validator(mode="after")
    def always_fails(self):
        assert False
    @model_validator(mode="after")
    def also_fails(self):
        raise ValueError("foo")
    a: int
    b: int
```

![Screenshot 2024-01-31 at 11 42 00 PM](https://github.com/pydantic/pydantic/assets/134332337/7d3409d8-3681-4795-9047-5193f3719adb)

Working on a slightly different approach similar to @forrest-bajbek also.

**Comment by ExcellentAmericanEagle** [comment]
Is there an official stance on that and best practice?

I want to show users validation errors on both model-level and field-level but:

If I have a **before model**-level validation error, it prevents showing field level validations.

Or if I have field level validations, it prevents raising **after model**-level validation errors unless all field level validations are fixed...

This is very frustrating there's nothing built in or an official example...

**Comment by hbutani** [comment]
FWIW, here is how I got `pydantic.ValidationError.from_exception_data` to work for 'value_error' errors(instead of defining a `PydanticCustomError`)
(thanks for the earlier responses for pointing all of us on how to do this...)

posting my function with example_error_list ; should be straightforward to repurpose for your situation.
(The 'ctx' param needed a 'error' entry)


```
from typing import Optional, Tuple, Sequence, TypeAlias
from pydantic import ValidationError
from pydantic_core import InitErrorDetails

FIELD_PATH_TYPE: TypeAlias = Tuple[str | int, ...]

example_error_list: list[dict[str, str | FIELD_PATH_TYPE]] = [{'loc': ('tables', 'name'), 'msg': 'Table tpcds.date not found in Knowledge Base', 'type': 'value_error'}]

def _raise_pydantic_validation_error(self,
                                       error_list: list[dict[str, str | FIELD_PATH_TYPE]]
                                       ) -> None:
      """
      Based on discussion in issue: https://github.com/pydantic/pydantic/discussions/7470
      and a lot of experimentation.

      :param error_list:
      :return:
      """

      if error_list:
          validation_errors = [
              InitErrorDetails(
                  input=self._star_schema_model,
                  ctx={'error': ValueError(err['msg'])},
                  loc=err['loc'],  # type: ignore
                  type=err['type'], # type: ignore
                  msg=err['msg'] # type: ignore
              )
              for err in error_list
          ]
          raise ValidationError.from_exception_data(title=StarSchemaModel.__name__,
                                                    line_errors=validation_errors
                                                    )
 ```
